api/src/DataPersister/Util/AbstractDataPersister.php

```diff
@@ -0,0 +1,105 @@
+<?php
+
+namespace App\DataPersister\Util;
+
+use ApiPlatform\Core\DataPersister\ContextAwareDataPersisterInterface;
+use App\Entity\BaseEntity;
+use InvalidArgumentException;
+
+abstract class AbstractDataPersister implements ContextAwareDataPersisterInterface {
+    /**
+     * @param CustomActionListener[]   $customActionListeners
+     * @param PropertyChangeListener[] $propertyChangeListeners
+     */
+    public function __construct(
+        private string $classname,
+        private DataPersisterObservable $dataPersisterObservable,
+        private array $customActionListeners = [],
```

Comment:
Andere Optionen statt der Registrierung via Konstruktor wären noch:
1. Registrierung nach dem parent-Konstruktor z.B. via `parent::subscribeBefore($action, $closure)` / `parent::subscribeAfter($action, $closure)`
2. Magic methods `beforeAccept`, `afterReject` etc. auf der konkreten DataPersister Klasse
3. Abstrakte und magic methods `beforeCreate`, `afterAccept` etc. auf der Listener-Klasse, dafür gar keine abstrakten `beforeCreate` etc. Methoden auf dem DataPersister

Der Vorteil deiner Variante wird ersichtlich, wenn alle DataPersisters unterschiedliche Dinge in ihren Hooks machen müssen, und wenn wir mehrere separate Listeners für dieselbe Aktion brauchen (was aber bei den Standard-Aktionen nicht unterstützt ist). Der Vorteil meiner Variante 1. zeigt sich, wenn z.B. bei vielen Events ein generisches (parametrisiertes) Mail versendet oder eine generische Notification irgendwohin gesendet werden muss. Magic methods sind Geschmackssache, können aber den Code unter Umständen viel lesbarer machen.

Reply:
Die Idee war gar nicht, die default actions (post, patch) zu überschreiben, sondern für diese ein striktes gerüst mit abstrakten methoden zu haben, das die default use cases abdeckt.

Mehrere Listener für dieselbe aktion hatte ich auch nicht wirklich im sinn.

Für so komisches zeug wie bei der Invitation war dann die frage, ob ich mit magic methods + convention arbeite, und mir den methodennamen zusammenbaue, oder user explizit subscriben muss.
Ich habe mich für die zweite Variante entschieden.

Falls wir bei vielen Actions von verschiedenen DataPersisters was ähnliches machen möchten, würde ich die funktionalität in injectbare klassen (z.b. Services) auslagern, und erst später z.b etwas a la EventBus machen, mit dem man sich auf Actions auf unterschiedlichen Entities subscriben kann.